### ISSUE TITLE
ICE on loop {|_: [_; continue]| {}}

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#51761]
The following code ICEs on `nightly`, `stable` and `beta`:
```rust
fn main() {
    loop {|_: [_; continue]| {}}
}
```

The ICE message is:

```
error: internal compiler error: librustc_mir/hair/cx/expr.rs:551: invalid loop id for continue: not inside loop scope
```

As of filing this bug, PR #51731 does not fix it. If I apply it, I'm getting:
```
error: internal compiler error: librustc_mir/build/scope.rs:552: no enclosing breakable scope found
  --> $DIR/issue-XXXX.rs:12:19
   |
LL |     loop {|_: [_; continue]| {}}
   |                   ^^^^^^^^
```

**Comment by varkor** [comment]
Counterintuitively, this should probably compile successfully. It's not surprising that it's not working, though: it's a really weird edge case. I'll look into it.

**Comment by varkor** [comment]
I feel perhaps control-flow should just not be allowed inside type-level constants like this. That seems like the most logical decision. In which case, we should just add a new error for this.

**Comment by rkruppe** [comment]
As I said, I believe this is a plain old "continue outside of loop" errors. The `continue` token is lexically within the `loop { ... }`, but it's not part of the loop body, so it's invalid without even considering that it's supposed to be a constant. It's morally the same as `loop { fn foo() { continue } }`.

**Comment by DutchGhost** [comment]
What about the `break` version?
Not sure if the fix is applied in the latest nightly, but this still ICE's:
https://github.com/rust-lang/rust/issues/51707

Rewritting it using a `loop {}` instead of `while`:

```Rust
fn main() {
    loop {
        |_: [_; break] | {}
    }
}
```

**Comment by est31** [comment]
@DutchGhost that's fixed by @varkor 's PR #51731 . The fix is not in nightly yet, not even merged. I've applied the PR locally in order to test whether the `loop {loop {|_: [_; continue]| {}} }` bug still exists.

It would probably make sense if @varkor would also fix the break version right away.
